README.md
# Telescope DNS

[![GitHub release (latest by date)](https://img.shields.io/github/v/release/wolf-joe/ts-dns)](https://github.com/wolf-joe/ts-dns/releases)
[![Go](https://github.com/wolf-joe/ts-dns/actions/workflows/go.yml/badge.svg)](https://github.com/wolf-joe/ts-dns/actions/workflows/go.yml)
[![codecov](https://codecov.io/gh/wolf-joe/ts-dns/branch/master/graph/badge.svg)](https://codecov.io/gh/wolf-joe/ts-dns)
[![Go Report Card](https://goreportcard.com/badge/github.com/wolf-joe/ts-dns)](https://goreportcard.com/report/github.com/wolf-joe/ts-dns)
![GitHub](https://img.shields.io/github/license/wolf-joe/ts-dns)

> 灵活快速的DNS分组转发器

## 设计目标
### 灵活解析
* 支持按ABP风格规则/`GFWList`对DNS请求进行分组
* 支持按CIDR对DNS请求进行重定向
* 支持DNS over UDP/TCP/TLS/HTTPS、socks5代理、ECS
* 支持将查询结果中的IPv4地址添加至IPSet
### 快速解析
* 支持并发请求上游DNS，选择最快响应
* 选择ping值最低的IPv4地址（tcp/icmp ping）
* 支持hosts/DNS缓存/屏蔽指定查询类型
* 支持热重载配置文件

## 解析流程
![arch.drawio.svg](doc/arch.drawio.svg)

```
查找hosts -> 查找缓存 -> 匹配规则 -> 指定group处理 -> 重定向 -> 设置缓存
```

## 使用说明

1. 在[Releases页面](https://github.com/wolf-joe/ts-dns/releases)下载对应系统和平台的压缩包；
2. 解压后按需求编辑配置文件`ts-dns.toml`（可选）并运行进程：
  ```shell
  # ./ts-dns -h  # 显示命令行帮助信息
  # ./ts-dns -c ts-dns.toml  # 指定配置文件名
  ./ts-dns
  kill -SIGHUP <PID> # 重载配置文件
  ```

## 配置示例

> 完整配置文件参见`ts-dns.full.toml`

1. 默认配置（`ts-dns.toml`），开箱即用
  ```toml
  listen = ":53"

  [groups]
    [groups.clean]
    dns = ["223.5.5.5", "114.114.114.114"]
    concurrent = true

    [groups.dirty]
    dns = [""] # 省略
    gfwlist_file = "gfwlist.txt"
  ```

2. 选择ping值最低的IPv4地址（启用时建议以root权限运行本程序）
  ```toml
  # ...
  [groups.clean]
    dns = ["223.5.5.5", "114.114.114.114"]
    fastest_v4 = true
  # ...
  ```

3. 指定hosts文件和自定义hosts
  ```toml
  # ...
  hosts_files = ["adaway.txt"]
  [hosts]
  "www.example.com" = "1.1.1.1"
  # ...
  ```

4. 使用socks5代理转发DNS请求
  ```toml
  # ...
    [groups.dirty]
    socks5 = "127.0.0.1:1080"
    # ...
  ```

5. 转发至上游DNS时默认附带指定ECS信息（暂不支持DOH）
  ```toml
  # ...
    [groups.clean]
    ecs = "1.2.4.0/24"
    # ...
  ```

6. 自定义域名分组
  ```toml
  # ...
    [groups.work]
    dns = ["10.1.1.1"]
    rules = ["company.com"]
    # ...
  ```

7. 动态添加IPSet记录（使用前请阅读`ts-dns.full.toml`对应说明）
  ```toml
  # ...
    [groups.dirty]
    ipset = "blocked"
    ipset_ttl = 86400
    # ...
  ```


## 未来规划

- [ ] 支持定期拉取最新gfwlist
- [ ] 支持http接口管理
- [ ] 降低gfwlist的匹配优先级
- [ ] DoT/GFWList域名解析自闭环

## 特别鸣谢
* [github.com/arloan/prdns](https://github.com/arloan/prdns)
* [github.com/gfwlist/gfwlist](https://github.com/gfwlist/gfwlist)


changelog.md
# v1.0.0

- [x] 从配置中移除`query_log`、`gfwlist`、`gfwlist_b64`项
- [x] 移除针对`dirty`、`clean`组的特殊逻辑
- [x] 支持为特定组指定`gfwlist`匹配策略、兜底匹配策略
- [x] 收到`SIGNUP`信号时重载配置文件
- [x] 支持非CNIP转发到指定组策略
- [X] 完全重构代码

cmd/config.go
package main

import (
	"github.com/BurntSushi/toml"
	log "github.com/Sirupsen/logrus"
	"github.com/janeczku/go-ipset/ipset"
	"github.com/wolf-joe/ts-dns/cache"
	"github.com/wolf-joe/ts-dns/hosts"
	"github.com/wolf-joe/ts-dns/inbound"
	"github.com/wolf-joe/ts-dns/matcher"
	"github.com/wolf-joe/ts-dns/outbound"
	"golang.org/x/net/proxy"
	"regexp"
	"strings"
	"time"
)

var config tomlStruct

type tomlStruct struct {
	Listen     string
	GFWList    string   `toml:"gfwlist"`
	CNIP       string   `toml:"cnip"`
	HostsFiles []string `toml:"hosts_files"`
	Hosts      map[string]string
	Cache      cacheStruct
	GroupMap   map[string]groupStruct `toml:"groups"`
}

type groupStruct struct {
	Socks5    string
	IPSetName string `toml:"ipset"`
	IPSetTTL  int    `toml:"ipset_ttl"`
	DNS       []string
	DoT       []string
	DoH       []string
	Rules     []string
}

type cacheStruct struct {
	Size   int
	MinTTL int `toml:"min_ttl"`
	MaxTTL int `toml:"max_ttl"`
}

// 从配置文件里读取ts-dns的配置并打包
func initHandler(filename string) (h *inbound.Handler) {
	var err error
	if _, err = toml.DecodeFile(filename, &config); err != nil {
		log.WithField("file", filename).Fatalf("read config error: %v", err)
	}
	// 默认配置
	if config.Listen == "" {
		config.Listen = ":53"
	}
	if config.GFWList == "" {
		config.GFWList = "gfwlist.txt"
	}
	if config.CNIP == "" {
		config.CNIP = "cnip.txt"
	}

	h = &inbound.Handler{Listen: config.Listen, GroupMap: map[string]*inbound.Group{}}
	// 读取gfwlist
	if h.GFWMatcher, err = matcher.NewABPByFile(config.GFWList, true); err != nil {
		log.WithField("file", config.GFWList).Fatalf("read gfwlist error: %v", err)
	}
	// 读取cnip
	if h.CNIP, err = cache.NewRamSetByFn(config.CNIP); err != nil {
		log.WithField("file", config.CNIP).Fatalf("read cnip error: %v", err)
	}
	// 读取Hosts列表
	var lines []string
	for hostname, ip := range config.Hosts {
		lines = append(lines, ip+" "+hostname)
	}
	if len(lines) > 0 {
		text := strings.Join(lines, "\n")
		h.HostsReaders = append(h.HostsReaders, hosts.NewTextReader(text))
	}
	// 读取Hosts文件列表。reloadTick为0代表不自动重载hosts文件
	for _, filename := range config.HostsFiles {
		if reader, err := hosts.NewFileReader(filename, 0); err != nil {
			log.WithField("file", filename).Warnf("read hosts error: %v", err)
		} else {
			h.HostsReaders = append(h.HostsReaders, reader)
		}
	}
	// 读取每个域名组的配置信息
	for groupName, groupConf := range config.GroupMap {
		// 读取socks5代理地址
		var dialer proxy.Dialer
		if groupConf.Socks5 != "" {
			dialer, _ = proxy.SOCKS5("tcp", groupConf.Socks5, nil, proxy.Direct)
		}
		// 为每个出站dns服务器创建对应Caller对象
		var callers []outbound.Caller
		for _, addr := range groupConf.DNS { // TCP/UDP服务器
			useTcp := false
			if strings.HasSuffix(addr, "/tcp") {
				addr, useTcp = addr[:len(addr)-4], true
			}
			if addr != "" {
				if !strings.Contains(addr, ":") {
					addr += ":53"
				}
				if useTcp {
					callers = append(callers, &outbound.TCPCaller{Address: addr, Dialer: dialer})
				} else {
					callers = append(callers, &outbound.UDPCaller{Address: addr, Dialer: dialer})
				}
			}
		}
		for _, addr := range groupConf.DoT { // dns over tls服务器，格式为ip:port@serverName
			var serverName string
			if arr := strings.Split(addr, "@"); len(arr) != 2 {
				continue
			} else {
				addr, serverName = arr[0], arr[1]
			}
			if addr != "" {
				if !strings.Contains(addr, ":") {
					addr += ":853"
				}
				if serverName != "" {
					callers = append(callers, outbound.NewTLSCaller(addr, dialer, serverName, false))
				}
			}
		}
		dohReg := regexp.MustCompile(`^https://.+/dns-query$`)
		for _, addr := range groupConf.DoH { // dns over https服务器，格式为https://domain/dns-query
			if dohReg.MatchString(addr) {
				callers = append(callers, &outbound.DoHCaller{Url: addr, Dialer: dialer})
			}
		}
		group := &inbound.Group{Callers: callers}
		// 读取匹配规则
		group.Matcher = matcher.NewABPByText(strings.Join(groupConf.Rules, "\n"))
		// 读取IPSet配置
		if groupConf.IPSetName != "" {
			if groupConf.IPSetTTL > 0 {
				group.IPSetTTL = groupConf.IPSetTTL
			}
			group.IPSet, err = ipset.New(groupConf.IPSetName, "hash:ip", &ipset.Params{})
			if err != nil {
				log.Fatalf("create ipset error: %v", err)
			}
		}
		h.GroupMap[groupName] = group
	}
	// 读取cache配置
	cacheSize, minTTL, maxTTL := 4096, time.Minute, 24*time.Hour
	if config.Cache.Size != 0 {
		cacheSize = config.Cache.Size
	}
	if config.Cache.MinTTL >= 0 {
		minTTL = time.Second * time.Duration(config.Cache.MinTTL)
	}
	if config.Cache.MaxTTL >= 0 {
		maxTTL = time.Second * time.Duration(config.Cache.MaxTTL)
	}
	if maxTTL < minTTL {
		maxTTL = minTTL
	}
	h.Cache = cache.NewDNSCache(cacheSize, minTTL, maxTTL)
	// 检测配置有效性
	if len(h.GroupMap) <= 0 || len(h.GroupMap["clean"].Callers) <= 0 || len(h.GroupMap["dirty"].Callers) <= 0 {
		log.Fatalf("dns of clean/dirty group cannot be empty")
	}
	return
}
